homeassistant/components/rest/binary_sensor.py

```diff
@@ -114,16 +125,70 @@ def is_on(self):
         """Return true if the binary sensor is on."""
         return self._is_on
 
+    @property
+    def extra_state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
     def _update_from_rest_data(self):
         """Update state from the rest data."""
-        if self.rest.data is None:
+        response = self.rest.data
+        _LOGGER.debug("Data fetched from resource: %s", response)
+
+        if response is None:
             self._is_on = False
 
-        response = self.rest.data
+        if self.rest.headers is not None:
+            # If the http request failed, headers will be None
+            content_type = self.rest.headers.get("content-type")
+
+            if content_type and (
+                content_type.startswith("text/xml")
+                or content_type.startswith("application/xml")
+                or content_type.startswith("application/xhtml+xml")
+                or content_type.startswith("application/rss+xml")
+            ):
+                try:
+                    response = json_dumps(xmltodict.parse(response))
+                    _LOGGER.debug("JSON converted from XML: %s", response)
+                except ExpatError:
+                    _LOGGER.warning(
+                        "REST xml result could not be parsed and converted to JSON"
+                    )
+                    _LOGGER.debug("Erroneous XML: %s", response)
+
+        if self._json_attrs:
+            self._attributes = {}
+            if response:
+                try:
+                    json_dict = json_loads(response)
+                    if self._json_attrs_path is not None:
+                        json_dict = jsonpath(json_dict, self._json_attrs_path)
+                    # jsonpath will always store the result in json_dict[0]
+                    # so the next line happens to work exactly as needed to
+                    # find the result
+                    if isinstance(json_dict, list):
+                        json_dict = json_dict[0]
+                    if isinstance(json_dict, dict):
+                        attrs = {
+                            k: json_dict[k] for k in self._json_attrs if k in json_dict
+                        }
+                        self._attributes = attrs
```

Comment:
you probably don't need `attrs`

Reply:
Do you mean to change it into:

```python
                    if isinstance(json_dict, dict):
                        self._attributes = {
                            k: json_dict[k] for k in self._json_attrs if k in json_dict
                        }
```

as `attrs` is not referenced anymore?

Reply:
Removed the uneccessary assignment